File: python/toy-problems/bubble_sort.py
# Failed attempt
def bubble(arr):
    n = len(arr)
    for j in range(0, n):
        sorted = False
        for i in range(0, n - 1):
            print(arr[i + 1])
            if arr[i] > arr[i + 1]:
                swap = arr[i]
                arr[i] = arr[i + 1]
                arr[i + 1] = swap
                sorted = True
                print(j, arr)
        if sorted == False:
            break

    return j, arr


arr_template = [33, 19, 97, 42, 38, 69, 73, 54, 1, 24, 25,
                43, 46, 75, 70, 9, 31, 49, 36, 97, 46, 52, 44, 51, 63, 25,
                39, 44, 14, 19, 89, 50, 100, 11, 40, 3, 14, 44, 38, 95, 9,
                58, 49, 4, 12, 56, 52, 25, 14, 27, 28, 18, 31, 80, 59, 84,
                70, 6, 76, 31, 100, 30, 81, 19, 54, 76, 60, 8, 42, 35, 90,
                9, 90, 97, 6, 8, 27, 47, 10, 55, 52, 50, 38, 79, 71, 81, 89,
                59, 37, 27, 82, 23, 51, 87, 27, 90, 38, 25, 27, 46]

print(bubble(arr_template))

File: javascript/toy-problems/bubble_sort.js
// Bubble Sort

// Since everybody hates chaos and loves sorted lists 
// we should implement some more sorting algorithms. 
// Your task is to implement a Bubble sort 
// (for some help look at https://en.wikipedia.org/wiki/Bubble_sort) 
// and return a list of snapshots after each change of the initial list.

// e.g.

// If the initial list would be l=[1,2,4,3] my algorithm rotates l[2] and l[3] and after that it adds [1,2,3,4] to the result, which is a list of snapshots.

const bubbleSort = (arr) => {
  const swap = (arry, idx1, idx2) => [arr[idx1], arr[idx2]] = [arr[idx2], arr[idx1]]
  let noSwap;

  for (let i = arr.length; i > 0; i--) {
    noSwap = true;
    for (let j = 0; j < i - 1; j++) {
      if (arr[j] > arr[j + 1]) {
        swap(arr, j, j + 1);
        noSwap = false
      }
    }
    if (noSwap) break;
  }

  return arr
};

const arr1 = [1,2,4,3] // should return [ [1,2,3,4] ]
const arr2 = [2,1,4,3] // should return [ [1,2,4,3], [1,2,3,4] ]
const arr3 = [1,2,3,4] // should return []

console.log(bubbleSort(arr1))
console.log(bubbleSort(arr2))
console.log(bubbleSort(arr3))